### PR TITLE
rustc_typeck: Allow reification from fn item to unsafe ptr

### PR DESCRIPTION
**Pull request by cramertj**
See https://github.com/rust-lang/rfcs/issues/1762.

I've never contributed to the compiler internals before-- apologies if I'm not going about this the right way.

### CODE REVIEW

File: src/librustc_typeck/check/coercion.rs [hunk 1]
```diff
@@ -196,6 +196,8 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
                 // Function items are coercible to any closure
                 // type; function pointers are not (that would
                 // require double indirection).
+                // Additionally, we permit coercin of function
```
Comment (Mark-Simulacrum) [inline src/librustc_typeck/check/coercion.rs]: nit: coercin -> coercion

File: src/librustc_typeck/check/coercion.rs [hunk 2]
```diff
@@ -504,6 +506,24 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
...
     }
 
+    fn coerce_from_safe_fn(&self,
+                           a: Ty<'tcx>,
+                           fn_ty_a: &'tcx ty::BareFnTy<'tcx>,
+                           b: Ty<'tcx>)
+                           -> CoerceResult<'tcx> {
+        if let ty::TyFnPtr(fn_ty_b) = b.sty {
+            match (fn_ty_a.unsafety, fn_ty_b.unsafety) {
+                (hir::Unsafety::Normal, hir::Unsafety::Unsafe) => {
```
Comment (eddyb) [inline src/librustc_typeck/check/coercion.rs]: This breaks coercing from an `unsafe fn` item to an `unsafe fn` pointer.

File: src/librustc_typeck/check/coercion.rs (hunk 1 above)
Comment (cramertj) [inline src/librustc_typeck/check/coercion.rs]: Good catch!

**Comment by brson** [comment]
This feature enhancement went landed insta-stable. I don't see any discussion of why that decision was made? Can somebody explain?

**Comment by cramertj** [comment]
@brson I opened up https://github.com/rust-lang/rfcs/issues/1762 for discussion, but my impression from  the comments there was that this was simply a missing implementation. The appropriate conversions already existed (`fn` item -> `fn` pointer and `fn` pointer -> `unsafe fn` pointer), but the current logic didn't allow them to be applied transitively. This commit just adds the missing transitive conversion.

However, if this requires more discussion, I'd be happy to revert and open a more formal RFC.

**Comment by petrochenkov** [comment]
@brson 
The relevant RFC is https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md and the tracking issue is https://github.com/rust-lang/rust/issues/18469#issuecomment-61578672.
This PR implements part of "add all transitive coercions" sub-item (which is also tracked by https://github.com/rust-lang/rust/issues/18602).

**Comment by nikomatsakis** [comment]
I agree that this feels more like a bugfix than a language change.

**Pull request merged**